src/WebJobs.Extensions.DurableTask/Listener/DurableTaskTargetScaler.cs

```diff
@@ -38,46 +39,54 @@ public DurableTaskTargetScaler(string functionId, DurableTaskMetricsProvider met
+
+                // compute activityWorkers: the number of workers we need to process all activity messages
+                var workItemQueueLength = metrics.WorkItemQueueLength;
+                double activityWorkers = Math.Ceiling(workItemQueueLength / (double)this.MaxConcurrentActivities);
+
+                var serializedControlQueueLengths = metrics.ControlQueueLengths;
+                var controlQueueLengths = JsonConvert.DeserializeObject<IReadOnlyList<int>>(serializedControlQueueLengths);
+
+                var controlQueueMessages = controlQueueLengths.Sum();
+                var activeControlQueues = controlQueueLengths.Count(x => x > 0);
+
+                // compute orchestratorWorkers: the number of workers we need to process all orchestrator messages.
+                // We bound this result to be no larger than the partition count
+                var upperBoundControlWorkers = Math.Ceiling(controlQueueMessages / (double)this.MaxConcurrentOrchestrators);
+                var orchestratorWorkers = Math.Min(activeControlQueues, upperBoundControlWorkers);
+
+                int numWorkersToRequest = (int)Math.Max(activityWorkers, orchestratorWorkers);
+                this.scaleResult.TargetWorkerCount = numWorkersToRequest;
+
+                // When running on ScaleController V3, ILogger logs are forwarded to the ScaleController's Kusto table.
+                // This works because this code does not execute in the Functions Host process, but in the ScaleController process,
+                // and the ScaleController is injecting it's own custom ILogger implementation that forwards logs to Kusto.
+                var metricsLog = $"Metrics: workItemQueueLength={workItemQueueLength}. controlQueueLengths={serializedControlQueueLengths}. " +
+                    $"maxConcurrentOrchestrators={this.MaxConcurrentOrchestrators}. maxConcurrentActivities={this.MaxConcurrentActivities}";
+                var scaleControllerLog = $"Target worker count for '{this.functionId}' is '{numWorkersToRequest}'. " +
+                    metricsLog;
+
+                // target worker count should never be negative
+                if (numWorkersToRequest < 0)
+                {
+                    throw new ArithmeticException("Number of workers to request cannot be negative");
+                }
+
+                this.logger.LogInformation(scaleControllerLog);
+                return this.scaleResult;
+            }
+            catch (Exception ex)
+            {
+                var metricsLog = $"Metrics: workItemQueueLength={metrics?.WorkItemQueueLength}. controlQueueLengths={metrics?.ControlQueueLengths}. " +
+                    $"maxConcurrentOrchestrators={this.MaxConcurrentOrchestrators}. maxConcurrentActivities={this.MaxConcurrentActivities}";
+                var errorLog = $"Error: target worker count for '{this.functionId}' resulted in exception." + metricsLog + $"Exception: {ex}";
+                throw new Exception(errorLog, ex);
```

Comment:
nit: we're effectively double-logging the exception here my including the same exception information in both `errorLog` and as an inner exception `ex`. This can be confusing for someone trying to debug scale controller logs.

Reply:
You're right, I think I was being overly verbose here. I've updated the code to remove the string representation of the error and instead just rely on the "innerException" parameter: https://github.com/Azure/azure-functions-durable-extension/pull/2462/commits/f5c6d3f1c14622f3383ca7055f71a82bb79c2a9e